src/boot/store.js

```diff
@@ -373,6 +385,19 @@ const replacer = function replacer(key, value) {
       [SERIALIZED_TYPE_FIELD_NAME_ESCAPED]: value[SERIALIZED_TYPE_FIELD_NAME],
     };
   }
+
+  // `origValue.toJSON` and `Object.getPrototypeOf(origValue)` fail
+  // if origValue is undefined or null.
+  if (origValue !== undefined && origValue !== null) {
+    // Don't forget to handle a value's `toJSON` method, if present, as
+    // described above.
+    invariant(typeof origValue.toJSON !== 'function', 'unexpected toJSON');
+
+    // If storing an interesting data type, don't forget to handle it
+    // here, and in `reviver`.
+    invariant(boringPrototypes.includes(Object.getPrototypeOf(origValue)), 'unexpected class');
```

Comment:
> I think we can do that by only going through this path when `typeof origValue === 'object' && origValue !== null`. That eliminates all but `Object.prototype` and `Array.prototype` as acceptable prototypes here.

Ah, right—makes sense, thanks.

> Hmm, or another thought that suggests: the whole `if`/`else if` chain, or the bulk of it, could perhaps turn into a `switch (Object.getPrototypeOf(origValue))`.

I think this might make sense. It would be slightly more rigid, in that it builds in the assumption that the distinguishing mark we're looking for is the identity of the first item in the prototype chain.

If the `AvatarURL` code had worked out slightly differently, and the three subclasses (`UploadedAvatarURL`, `FallbackAvatarURL`, `GravatarURL`) just inherited their `serialize` and `deserialize` methods from their `AvatarURL` superclass, without overriding them, I could see it being convenient to do an `if (value instanceof AvatarURL)` instead of `if (value instanceof GravatarURL)` and friends; this kind of simplification wouldn't work with the proposed switch/case.

Also, we use an `if (Immutable.Map.isMap(origValue))`—we'd have to double-check our expectation that `if (origValue instanceof Immutable.Map)` would be equivalent before refactoring that to the switch/case.

Reply:
> I think this might make sense. It would be slightly more rigid, in that it builds in the assumption that the distinguishing mark we're looking for is the identity of the first item in the prototype chain.
> 
> If the `AvatarURL` code had worked out slightly differently, and the three subclasses (`UploadedAvatarURL`, `FallbackAvatarURL`, `GravatarURL`) just inherited their `serialize` and `deserialize` methods from their `AvatarURL` superclass, without overriding them, I could see it being convenient to do an `if (value instanceof AvatarURL)` instead of `if (value instanceof GravatarURL)` and friends; this kind of simplification wouldn't work with the proposed switch/case.

True!

There'd still be a pretty simple way to handle that, though:
```js
  case GravatarURL:
  case UploadedAvatarURL:
  case FallbackAvatarURL:
    return { data: AvatarURL.serialize(value), [SERIALIZED_TYPE_FIELD_NAME]: `AvatarURL` };
```

It involves just a bit more explicit detail about the subclasses than `value instanceof AvatarURL` would. But I think that isn't too bad of a thing.

> Also, we use an `if (Immutable.Map.isMap(origValue))`—we'd have to double-check our expectation that `if (origValue instanceof Immutable.Map)` would be equivalent before refactoring that to the switch/case.

Yes. From the implementation, I'm pretty sure those two are equivalent. I think the `isFoo` form comes from wanting some overlapping categories like `isKeyed` and `isIndexed`, which the `instanceof` approach would only be able to model in a language with multiple inheritance. (It seems to originate, with the same implementation strategy used for `isMap` now, in immutable-js/immutable-js@921b937e6ac2bf431f968f6a25dcffbbf49bc08f.)

Alternatively, if there were a few things (like `Immutable.Map` in a slightly different possible world) which didn't fit the `switch (prototype)`, we could always handle them as plain conditions the same as before -- just in the `default` case of the `switch`, or after the `switch`.

Reply:
Cool, makes sense—I just opened #4362 for this.